rust/src/traits/samplers/bernoulli/mod.tex

```diff
@@ -0,0 +1,327 @@
+\newtheorem{question}{Question}
+
+\title{Sampler Proofs for OpenDP: \texttt{sample\_bernoulli}}
+\author{Vicki Xu, Hanwen Zhang, Zachary Ratliff}
+\date{Summer 2022}
+\begin{document}
+
+\maketitle
+\tableofcontents
+
+\begin{tcolorbox}
+\begin{warning}
+The \texttt{sample\_bernoulli} function relies on a cryptographic pseudorandom number generator (CSPRNG) provided by the \href{https://www.openssl.org/docs/man1.1.1/man3/RAND_bytes.html}{OpenSSL library}. As a result, measurements that use $\texttt{sample\_bernoulli}$ as their source of randomness will at most satisfy \emph{computational} differential privacy parameterized by a security parameter $\lambda$ (where the OpenSSL security parameter defaults to $\lambda = 256$ bits). 
+\end{warning}
+\end{tcolorbox}
+
+\begin{tcolorbox}
+\begin{warning}[Code is not constant-time]
+ \texttt{sample\_bernoulli} takes in a boolean \texttt{constant\_time} parameter to protect against timing attacks on the Bernoulli sampling procedure. However, the current implementation does not guard against other types of timing side-channels that can break differential privacy, e.g., non-constant time code execution due to branching.
+\end{warning}
+\end{tcolorbox}
+
+\section{Algorithm Implementation}
+\subsection{Code in Rust}
+
+The current OpenDP library contains the \texttt{sample\_bernoulli} function, which returns a draw from a Bernoulli distribution with parameter \texttt{prob}. This is defined in lines 59-92 of the file \texttt{mod.rs} in the Git repository \url{https://github.com/opendp/opendp/blob/main/rust/src/traits/samplers/bernoulli/mod.rs#L59-L92}. 
+
+\texttt{sample\_bernoulli} accepts a parameter \texttt{prob} of generic type \texttt{T} and another parameter \texttt{constant\_time} of type \texttt{bool}. As aforementioned, \texttt{prob} is the parameter for the draw from the distribution. \texttt{constant\_time} is an option for \texttt{sample\_bernoulli} to accomplish this draw in near-constant time, which mitigates timing attacks. This proof will not consider non-constant time.
+
+\vicki{high-level walkthrough - should this be a separate section?} 
+\texttt{sample\_bernoulli} makes use of \texttt{sample\_geometric\_buffer} and \texttt{fill\_bytes}. As defined in line \ref{line:fillbytes} of the Python pseudocode below, \texttt{fill\_bytes} takes in an empty array \texttt{buf} and fills it with random bytes generated with a cryptographically-secure pseudorandom number generator in the OpenSSL library. As defined in line \ref{line:geombuffer} of the pseudocode, \texttt{sample\_geometric\_buffer} declares an byte buffer \texttt{buf} with \texttt{B} elements that is then filled using the \texttt{fill\_bytes} function. Then out of the nonzero elements of \texttt{buf}, \texttt{sample\_geometric\_buffer} treats \texttt{buf} as a string of bits and returns the zero-indexed position of the first $1$ (as in line \ref{line:indexcmp}) or \texttt{None} if no $1$'s are found in this string of bits. 
```

Comment:
Minor nit, if you split the text into more lines, then review comments can be more specific.

The impl of SampleBernoulli for bool only directly depends on the correctness of one other sampler, `sample_geometric_buffer`. I realize you are waiting to split the proofs apart, though.

The `B` is also stale-- the `B` const generic argument became a normal argument in the effort to get rid of explicit references to concrete types in the code (blanket impls).